Source/JavaScriptCore/runtime/AtomicsObject.cpp

```diff
@@ -455,8 +456,52 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
         return { };
     }
 
-    AtomicsWaitValidation validation = WTF::atomicLoad(ptr) == expectedValue ? AtomicsWaitValidation::Pass : AtomicsWaitValidation::Fail;
-    return WaiterListManager::singleton().wait(globalObject, vm, ptr, validation, timeout, type);
+    switch (type) {
+    case AtomicsWaitType::Sync: {
+        auto list = WaiterListManager::singleton().getLockedWaiterList(ptr);
+
+        if (WTF::atomicLoad(ptr) != expectedValue)
+            return vm.smallStrings.notEqualString();
+
+        const auto waitTimedOut = list.waitSync(vm, timeout);
+        if (waitTimedOut)
+            return vm.smallStrings.timedOutString();
+        return vm.smallStrings.okString();
+    }
+
+    case AtomicsWaitType::Async: {
+        JSObject* object = constructEmptyObject(globalObject);
+        bool isAsync = false;
+        JSValue value;
+
+        if (!timeout)
+            value = vm.smallStrings.timedOutString();
+        else {
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
+
+            {
+                auto list = WaiterListManager::singleton().getLockedWaiterList(ptr);
+                if (WTF::atomicLoad(ptr) != expectedValue)
+                    value = vm.smallStrings.notEqualString();
+                else if (!timeout)
+                    value = vm.smallStrings.timedOutString();
+                else {
+                    isAsync = true;
+                    list.waitAsync(WTFMove(waiter), timeout);
+                    value = promise;
+                }
+            }
+        }
+
+        object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
+        object->putDirect(vm, vm.propertyNames->value, value);
+        return object;
+    }
+
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
```

Comment:
I think we should move these code in `WaiterListManager`. Extracting these part from `WaiterListManager` is making lock management harder (e.g., `LockedWaiterList` should not have Locker field.)